src/audio/simulated-host-zephyr.c

```diff
@@ -0,0 +1,1142 @@
+		dma_release_channel(hd->dma->z_dev, hd->chan->index);
+		hd->chan = NULL;
+	}
+
+	/* free all DMA elements */
+	dma_sg_free(&hd->host.elem_array);
+	dma_sg_free(&hd->local.elem_array);
+	dma_sg_free(&hd->config.elem_array);
+
+	/* free DMA buffer */
+	if (hd->dma_buffer) {
+		buffer_free(hd->dma_buffer);
+		hd->dma_buffer = NULL;
+	}
+
+	/* reset buffer pointers */
+	hd->local_pos = 0;
+	hd->report_pos = 0;
+	hd->total_data_processed = 0;
+
+	hd->copy_type = COMP_COPY_NORMAL;
+	hd->source = NULL;
+	hd->sink = NULL;
+}
+
+static int host_reset(struct comp_dev *dev)
+{
+	struct host_data *hd = comp_get_drvdata(dev);
+
+	comp_dbg(dev, "host_reset()");
+
+	host_common_reset(hd, dev->state);
+	dev->state = COMP_STATE_READY;
+
+	return 0;
+}
+
+static int host_copy(struct comp_dev *dev)
+{
+	struct host_data *hd = comp_get_drvdata(dev);
+
+	if (dev->state != COMP_STATE_ACTIVE)
+		return 0;
+
+	return host_common_copy(hd, dev, host_dma_cb);
+}
+
+static int host_get_attribute(struct comp_dev *dev, uint32_t type,
+			      void *value)
+{
+	struct host_data *hd = comp_get_drvdata(dev);
+
+	switch (type) {
+	case COMP_ATTR_COPY_TYPE:
+		*(enum comp_copy_type *)value = hd->copy_type;
+		break;
+	case COMP_ATTR_COPY_DIR:
+		*(uint32_t *)value = hd->ipc_host.direction;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int host_set_attribute(struct comp_dev *dev, uint32_t type,
+			      void *value)
+{
+	struct host_data *hd = comp_get_drvdata(dev);
+
+	switch (type) {
+	case COMP_ATTR_COPY_TYPE:
+		hd->copy_type = *(enum comp_copy_type *)value;
+		break;
+	case COMP_ATTR_HOST_BUFFER:
+		hd->host.elem_array = *(struct dma_sg_elem_array *)value;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static uint64_t host_get_processed_data(struct comp_dev *dev, uint32_t stream_no, bool input)
+{
+	struct host_data *hd = comp_get_drvdata(dev);
+	uint64_t ret = 0;
+	bool source = dev->direction == SOF_IPC_STREAM_PLAYBACK;
+
+	/* Return value only if direction and stream number match.
+	 * The host supports only one stream.
+	 */
+	if (stream_no == 0 && source == input)
+		ret = hd->total_data_processed;
+
+	return ret;
+}
+
+static const struct comp_driver comp_host = {
+	.type	= SOF_COMP_HOST,
+	.uid	= SOF_RT_UUID(host_uuid),
+	.tctx	= &host_tr,
+	.ops	= {
+		.create				= host_new,
+		.free				= host_free,
+		.params				= host_params,
+		.reset				= host_reset,
+		.trigger			= host_trigger,
+		.copy				= host_copy,
+		.prepare			= host_prepare,
+		.position			= host_position,
+		.get_attribute			= host_get_attribute,
+		.set_attribute			= host_set_attribute,
+		.get_total_data_processed	= host_get_processed_data,
+	},
+};
+
+static SHARED_DATA struct comp_driver_info comp_host_info = {
+	.drv = &comp_host,
+};
+
+UT_STATIC void sys_comp_host_init(void)
+{
+	comp_register(platform_shared_get(&comp_host_info,
+					  sizeof(comp_host_info)));
+}
+
+DECLARE_MODULE(sys_comp_host_init);
+SOF_MODULE_INIT(host, sys_comp_host_init);
```

Comment:
this and the dai version look like a lot of copy-paste, is this what we want to do? The code of the original source file will evolve over time so it's a maintenance nightmare to keep two separate files aligned by parts and different when they need to be different.

Code management/maintainability is hard.

Reply:
Good, question. This way I would remove the disabled code parts. Though I didn't yet start to work with host so it's now just a duplicated. Other way would be adding abstraction to zephyr DAI and host and use stub functions for DMA and HW related for the simulated version.